### PR TITLE
Add SASL_SSL for Kafka Auth

### PR DESCRIPTION
**Pull request by cliveseldon**
- Updates modelgateway, pipelinegateway and dataflow to handle new SASL passowrd settings
- Updates components to allow password download from Secret or file
- Update Ansible and Helm charts to allow SASL SCRAM option. For Strimzi you can't use this with Kraft at present.
- Updates docs 
- Update CLI to handle config for SASL_SSL
- Update CLI to allow namespace to be set in Kafka config (as its used for topic names)

### CODE REVIEW

File: components/tls/pkg/password/store_test.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+	}
+	_, err = clientset.CoreV1().Secrets(sec.Namespace).Update(context.Background(), &sec2, metav1.UpdateOptions{})
+	g.Expect(err).To(BeNil())
+	time.Sleep(time.Millisecond * 500)
+	passwordLoaded = ps.GetPassword()
+	g.Expect(passwordLoaded).To(Equal(string(password)))
+	ps.Stop()
+}
+
+func TestWatchFolderCertificate(t *testing.T) {
```
Comment (sakoush) [inline components/tls/pkg/password/store_test.go:54]: is the test name representative?

File: operator/pkg/cli/kafka.go [hunk 2]
```diff
@@ -86,11 +88,30 @@ func NewKafkaClient(kafkaBroker string, kafkaBrokerIsSet bool, schedulerHost str
...
+
 	if config.Kafka != nil {
-		consumerConfig["security.protocol"] = "ssl"
-		consumerConfig["ssl.ca.location"] = config.Kafka.CaPath
-		consumerConfig["ssl.key.location"] = config.Kafka.KeyPath
-		consumerConfig["ssl.certificate.location"] = config.Kafka.CrtPath
+		if config.Kafka.Namespace != "" {
+			namespace = config.Kafka.Namespace
+		}
+		if config.Kafka.Protocol == "ssl" {
```
Comment (sakoush) [inline operator/pkg/cli/kafka.go:97]: nit: should we have an enum for ` config.Kafka.Protocol`?

File: scheduler/data-flow/src/main/kotlin/io/seldon/dataflow/kafka/Configuration.kt [hunk 3]
```diff
@@ -98,6 +115,20 @@ fun getKafkaProperties(params: KafkaStreamsParams): KafkaProperties {
...
+            }
+            val trustStoreConfig = Provider.trustStoreFromCertificates(params.security.certConfig)
+            this[SslConfigs.SSL_TRUSTSTORE_LOCATION_CONFIG] = trustStoreConfig.trustStoreLocation
+            this[SslConfigs.SSL_TRUSTSTORE_PASSWORD_CONFIG] = trustStoreConfig.trustStorePassword
+            val password = K8sPasswordSecretsProvider.downloadPasswordFromSecret(params.security.saslConfig)
+            this[SaslConfigs.SASL_MECHANISM] = "SCRAM-SHA-512"
+            val jaasTemplate =
+                "org.apache.kafka.common.security.scram.ScramLoginModule required username=\"%s\" password=\"%s\";"
+            val jaasCfg = java.lang.String.format(jaasTemplate, params.security.saslConfig.username, password)
+            this[SaslConfigs.SASL_JAAS_CONFIG]= jaasCfg
```
Comment (sakoush) [inline scheduler/data-flow/src/main/kotlin/io/seldon/dataflow/kafka/Configuration.kt:131]: this looks identical to the code block above, could we extract as helper func?

File: scheduler/pkg/kafka/config/tls.go [hunk 4]
```diff
@@ -18,39 +18,75 @@ package config
...
+		return err
+	}
+	username, found := util.GetEnv(EnvKafkaClientPrefix, EnvSASLUsernameSuffix)
+	if !found {
+		username = DefaultSASLUsername
+	}
+	config["sasl.username"] = username
+	config["sasl.password"] = ps.GetPassword()
+	// issue is that ca.pem does not work with multiple certificiates defined
+	// see https://github.com/confluentinc/confluent-kafka-go/issues/827 (Fixed but not yet released)
```
Comment (sakoush) [inline scheduler/pkg/kafka/config/tls.go:67]: apparently this is now fixed as per [comment](https://github.com/confluentinc/confluent-kafka-go/issues/827#issuecomment-1303413207)?

File: scheduler/pkg/kafka/config/tls.go [hunk 5]
```diff
@@ -18,39 +18,75 @@ package config
...
+	cs, err := tls.NewCertificateStore(tls.Prefix(EnvKafkaClientPrefix),
+		tls.ValidationPrefix(EnvKafkaBrokerPrefix))
+	if err != nil {
+		return err
+	}
+	cert := cs.GetCertificate()
+	caCert := cs.GetValidationCertificate()
+	config["security.protocol"] = "ssl"
+	// issue is that ca.pem does not work with multiple certificiates defined
+	// see https://github.com/confluentinc/confluent-kafka-go/issues/827 (Fixed but not yet released)
```
Comment (sakoush) [inline scheduler/pkg/kafka/config/tls.go:82]: ditto

File: scheduler/pkg/kafka/config/tls.go [hunk 6]
```diff
@@ -18,39 +18,75 @@ package config
...
-			tls.ValidationPrefix(EnvKafkaBrokerPrefix))
-		if err != nil {
-			return err
-		}
+	switch protocol {
+	case tls.SecurityProtocolSSL:
+		return setupTLSAuthentication(config)
+	case tls.SecurityProtocolSASLSSL:
+		return setupSASLSSLAuthentication(config)
+	case tls.SecurityProtocolPlaintxt:
```
Comment (sakoush) [inline scheduler/pkg/kafka/config/tls.go:41]: should there be an option for SASL plaintext?

**Review by sakoush — COMMENTED** [review]
I left some clarification comments.

File: components/tls/pkg/password/file.go [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+	"os"
+	"sync"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+	"istio.io/pkg/filewatcher"
+)
+
+type PasswordFolderHandler struct {
+	prefix           string
```
[the remark this answers is not in the document]
Comment (cliveseldon) [reply components/tls/pkg/password/file.go:15]: will remove

File: components/tls/pkg/password/k8s_secret.go [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+}
+
+func (s *PasswordSecretHandler) GetPassword() string {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+	return s.password
+}
+
+func (s *PasswordSecretHandler) Stop() {
+	close(s.stopper)
```
[the remark this answers is not in the document]
Comment (cliveseldon) [reply components/tls/pkg/password/k8s_secret.go:59]: As these are library components I think its up to the caller to ensure this unless you have a better suggestion?

File: components/tls/pkg/password/k8s_secret.go [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+		}
+	}
+}
+
+func (s *PasswordSecretHandler) onDelete(obj interface{}) {
+	logger := s.logger.WithField("func", "onDelete")
+	secret := obj.(*corev1.Secret)
+	if secret.Name == s.secretName {
+		logger.Warnf("Secret %s deleted", secret.Name)
+	}
```
[the remark this answers is not in the document]
Comment (cliveseldon) [reply components/tls/pkg/password/k8s_secret.go:105]: Yes at present we don't do anything that warn the secret being watched is nolonger existing.

File: components/tls/pkg/password/k8s_secret_test.go [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+	password = []byte("newpassword")
+	sec2 := v1.Secret{
+		ObjectMeta: metav1.ObjectMeta{Name: secName, Namespace: "default"},
+		Data: map[string][]byte{
+			"password": password,
+		},
+	}
+	_, err = clientset.CoreV1().Secrets(sec.Namespace).Update(context.Background(), &sec2, metav1.UpdateOptions{})
+	g.Expect(err).To(BeNil())
+	time.Sleep(time.Millisecond * 500)
```
[the remark this answers is not in the document]
Comment (cliveseldon) [reply components/tls/pkg/password/k8s_secret_test.go:50]: to provide enough time for watcher to signal change and do the update

File: components/tls/pkg/password/store_test.go (hunk 1 above)
Comment (cliveseldon) [reply components/tls/pkg/password/store_test.go:54]: will change

File: operator/pkg/cli/kafka.go (hunk 2 above)
Comment (cliveseldon) [reply operator/pkg/cli/kafka.go:97]: I'll change to some constants for now. These would need to match the available values used in kafka.

File: scheduler/pkg/kafka/config/tls.go (hunk 4 above)
Comment (cliveseldon) [reply scheduler/pkg/kafka/config/tls.go:67]: Maybe leave to a future PR and upgrade of librdkafka we use.

File: scheduler/pkg/kafka/config/tls.go (hunk 5 above)
Comment (cliveseldon) [reply scheduler/pkg/kafka/config/tls.go:82]: same as above

File: scheduler/pkg/kafka/config/tls.go (hunk 6 above)
Comment (cliveseldon) [reply scheduler/pkg/kafka/config/tls.go:41]: Strimzi doesn't support this. We would need to find a Kafka provider that does and update here and in dataflow engine. Will add a comment.

**Review by sakoush — APPROVED** [review]
LGTM

**Pull request merged**